plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/ConfigSnapshotFactory.java

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright (c) 2020 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.xtext.workspace;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.n4js.utils.URIUtils;
+import org.eclipse.xtext.workspace.ISourceFolder;
+
+import com.google.common.collect.ImmutableBiMap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.inject.Inject;
+
+/**
+ * Factory for creating snapshots
+ */
+@SuppressWarnings("restriction")
+public class ConfigSnapshotFactory {
+
+	/** Provider to create {@link BuildOrderInfo} instances */
+	@Inject
+	protected BuildOrderFactory buildOrderFactory;
+
+	/** Creates an instance of {@link WorkspaceConfigSnapshot} */
+	public WorkspaceConfigSnapshot createWorkspaceConfigSnapshot(XIWorkspaceConfig workspaceConfig) {
+		List<? extends XIProjectConfig> projects = new ArrayList<>(workspaceConfig.getProjects());
+		List<ProjectConfigSnapshot> projectSnapshots = Lists.transform(projects, this::createProjectConfigSnapshot);
+
+		return createWorkspaceConfigSnapshot(workspaceConfig.getPath(), projectSnapshots);
+	}
+
+	/** Creates an instance of {@link WorkspaceConfigSnapshot} */
+	public WorkspaceConfigSnapshot createWorkspaceConfigSnapshot(URI path,
+			Iterable<? extends ProjectConfigSnapshot> projects) {
+
+		Map<String, ProjectConfigSnapshot> lookupName2Project = new HashMap<>();
+		Map<URI, ProjectConfigSnapshot> lookupProjectPath2Project = new HashMap<>();
+		Map<URI, ProjectConfigSnapshot> lookupSourceFolderPath2Project = new HashMap<>();
+		updateWorkspaceConfigSnapshotLookupMaps(lookupName2Project, lookupProjectPath2Project,
+				lookupSourceFolderPath2Project, projects, Collections.emptyList());
+
+		return createWorkspaceConfigSnapshot(path,
+				ImmutableBiMap.copyOf(lookupName2Project),
+				ImmutableMap.copyOf(lookupProjectPath2Project),
+				ImmutableMap.copyOf(lookupSourceFolderPath2Project));
+	}
+
+	/** Creates an instance of {@link WorkspaceConfigSnapshot} */
+	public WorkspaceConfigSnapshot createWorkspaceConfigSnapshot(WorkspaceConfigSnapshot snapshot) {
+
+		return createWorkspaceConfigSnapshot(snapshot.path,
+				snapshot.name2Project, snapshot.projectPath2Project, snapshot.sourceFolderPath2Project);
+	}
+
+	/** Creates an instance of {@link WorkspaceConfigSnapshot} */
+	public WorkspaceConfigSnapshot createWorkspaceConfigSnapshot(URI path) {
```

Comment:
Some clients are invoking this method with a path argument of `null`. Please add a comment that `path` may be null to API doc of this method and/or other methods in this class.

Reply:
Actually misleading since a null argument results in invalid instances. The only use case is inside N4JS in a temporary solution until the Eclipse product is removed. See comment at class `NullXWorkspaceConfigSnapshotProvider`.